fontforge/splineutil.c

```diff
@@ -7953,3 +7953,19 @@ bigreal DistanceBetweenPoints( BasePoint *p1, BasePoint *p2 )
     t = sqrt( t );
     return t;
 }
+
+int SplineSetNumberPoints(SplineSet *ss, int order2) {
+    int pnum=0;
+    if ( order2 ) {
+        return SSTtfNumberPoints(ss);
+    } else {
+        for ( SplinePoint *sp = ss->first; ; ) {
+            pnum++;
+            if ( sp->next==NULL ) break;
+            if ( !sp->nonextcp || !sp->next->to->noprevcp ) pnum += 2;
```

Comment:
I don't understand this -- if you have one cubic contour with eight on-curve points and fourteen off-curve points, and another cubic contour with 20 on-curve points and two off-curve points, are they somehow interpolation-compatible? I would have thought interpolation mostly cares about on-curve points and can fudge transitioning lines to curves by considering missing off-curve points to be co-located.

Reply:
Yes, they are compatible. The distinction does not matter to the interpolator, it only considers points, not whether they are on or off curve.

Reply:
Do the results make sense? I take it the idea with this PR is to add what amounts to an "only do this if it makes sense" flag. We might want to differentiate between cases the algorithm can operate on and cases with a chance of having a good result.

Reply:
Sure, that makes sense to me. I'll modify it to count off- and on-curve points separately, as we are calling this "strict" mode in the API so it makes sense that that would be a strict interpretation.

Reply:
I'm transitioning from reviewing to spit-balling here. 

Even assuming a relatively boring interpolator that is not trying to help the user *much* I think having the number of on-curve points match is what should matter. If a segment with control points is being matched up with a line there are some pretty simple heuristics the code can use for the line, for example (considering cubics):

1. Add co-linear control points 1/3 and 2/3 of the way along the line, then interpolate. 
2. (Getting a bit more sophisticated) Look at the distance of each control point on the curved segment compared with the distance between the on-curve points and add co-linear control points to the line at corresponding ratios, then interpolate. 

Now, whether FontForge's interpolation code can do this is another matter.

Reply:
It can't. The standard also rejects such heuristics. Given that our (my) ultimate goal with adding this is to help people produce CFF2 variable fonts, I disagree with your suggestion that I should ignore the number of off curve points. Improving FontForge's interpolator does nothing to improve the interpolator actually in use by e.g. FreeType.

Reply:
That's fine but I think it means that you can't take this approach at all. Knowing the number of off-curve points is pretty meaningless if you don't know how they are distributed among the on-curve points (isn't it?). Unless you want to do something exotic you'll probably need to iterate through the contour pairs to make sure they match up.

Incidentally, what does the standard say about co-located control points? The current code in the PR seems to be mapping the single point case onto one in which the other control point is co-located with the on-curve point. That's probably fine, and FontForge probably shouldn't be marking those `no?cp` anyway (see #4125) but the spec might have something to say about it if it prohibits lines from getting the same treatment.

(An example of "exotic" would be forming a bit string where 0 means a line and 1 means a curve and returning its checksum.)

Reply:
If you'd rather chat about this in some other setting sometime I'm happy to do that too. The other review issues are pretty ordinary while this one is close to the heart of the feature.

Reply:
Still got time? I'm in IRC

Reply:
By the way, for future reference, just joining IRC (#fontforge) and saying "copypaste" will alert me if I'm at PC :-)

My client is more relaxed for other channels where people usually say my name on accident (e.g. "copypaste this into a terminal") but I watch #fontforge very closely